codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -902,32 +918,45 @@ public void beginRead() {
                 if (readInProgress || !isActive()) {
                     return;
                 }
+                readInProgress = true;
 
                 final RecvByteBufAllocator.Handle allocHandle = unsafe().recvBufAllocHandle();
                 allocHandle.reset(config());
-                if (inboundBuffer == null) {
+                if (inboundBuffer == null || inboundBuffer.isEmpty()) {
                     if (closePending) {
                         unsafe.closeForcibly();
-                    } else {
-                        readInProgress = true;
                     }
                     return;
                 }
 
+                // We have already checked that the queue is not empty, so before this value is used it will always be
+                // set by allocHandle.continueReading().
+                boolean continueReading;
                 do {
                     Object m = inboundBuffer.poll();
                     if (m == null) {
-                        if (closePending) {
-                            pipeline().fireChannelReadComplete();
-                            unsafe.closeForcibly();
-                        }
-                        return;
+                        continueReading = false;
+                        break;
                     }
                     doRead0((Http2Frame) m, allocHandle);
-                } while (allocHandle.continueReading());
+                } while (continueReading = allocHandle.continueReading());
 
-                allocHandle.readComplete();
-                pipeline().fireChannelReadComplete();
+                if (continueReading && parentReadInProgress) {
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) Change this condition so that it does not always evaluate to "false" [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS2583)